### DISCUSSION TITLE
How to propagate subtractions from one Part into another?

### DISCUSSION THREAD

**Discussion by mbugert** [gumyr/build123d#665] [Q&A]
Hi there!
I wonder -- what is an idiomatic solution for propagating subtractions from one part into another?

As an example, consider modeling a 3D printed part that should be screwed onto a desk. The screws require pre-drilling, so the model of the desk needs to have holes. The location of the screws are dictated by the 3D printed part. The question is how to best propagate the location of the screws (which could depend on lengthy computations we want to hide to the outside) from the printed model into the desk model.

This is my primitive current solution:
```python
from build123d import *
from ocp_vscode import show_object

bolt_d = 4
clearance = 0.4

def make_hole():
    return Hole(radius=bolt_d/2 + clearance, depth=999)

def make_bolt():
    return Cylinder(radius=bolt_d/2, height=16)

def make_printed_part() -> tuple:
    locations = GridLocations(x_spacing=8, y_spacing=0, x_count=2, y_count=1)
    part = Box(15, 15, 3, align=(Align.CENTER, Align.CENTER, Align.MIN))
    part -= locations * make_hole()
    printed_part = Compound(color="grey", children=[part])
    return printed_part, locations

def make_desk(locations) -> Compound:
    desk = Box(30, 30, 5, align=(Align.CENTER, Align.CENTER, Align.MAX))
    desk -= locations * make_hole()
    return Compound(color="BurlyWood", children=[desk])

def make_assembly() -> Compound:
    printed_part, locations = make_printed_part()
    desk = make_desk(locations)
    bolts = locations * make_bolt()
    return Compound(children=[desk, printed_part, *bolts])

assy = make_assembly()
show_object(assy)
```
![print_desk_bolts](https://github.com/user-attachments/assets/e11a8155-6ffa-4d17-802d-94a7f298a0ce)

This "works", but if `printed_part` is moved, one needs to move `locations` along manually.

I wonder if there is a way to define a Compound that contains subtractions that are applied to other parts later. My crude idea was to define
```python
holes = locations * make_hole()
printed_part = Compound(color="grey", children=[part - holes, *holes])
```
followed by
```python
printed_part_subtractions = [e for e in printed_part.children if hasattr(e, "mode") and e.mode == Mode.SUBTRACT]
desk -= printed_part_subtractions
```
Apart from dangerously mixing builder and algebra mode features, the problem with this approach is that `printed_part` contains the holes as added geometry, unless one removes them again from `.children` before use.

From my understanding, RigidJoints will not help here, since they only handle parts orientation, not parts geometry.

So, what would be a more idiomatic way of modeling such a scenario?

**Comment by gumyr** [comment]
Good question.

Does the following work for you?

```py
from build123d import *
from ocp_vscode import show_all, show
from bd_warehouse.fastener import SocketHeadCapScrew, ClearanceHole

m3 = SocketHeadCapScrew("M3-0.5", 20)

with BuildPart() as part1:
    Box(60, 60, 20)
    with Locations(part1.faces().sort_by(Axis.Z)[-1]):
        with GridLocations(50, 50, 2, 2) as top_holes:
            ClearanceHole(m3)

with BuildPart() as part2:
    Box(200, 200, 200)
    with Locations(part2.faces().sort_by(Axis.X)[-1]):
        with Locations(*m3.hole_locations) as p2_holes:
            h = ClearanceHole(m3, counter_sunk=False)

show(pack([part1.part, part2.part], 10))
print(p2_holes.locations)
```
![image](https://github.com/user-attachments/assets/8efbae6e-4f7c-4fa0-bd39-3782ddeaab76)

The `ClearanceHole` (and other custom hole classes) stores the location of the holes created for a given fastener within that fastener instance so in this case `m3.hole_locations` is a list of the four locations in `part1` that have clearance holes.  The `m3.hole_locations` can then be used in `part2` to place the group of holes, in this case on the side of `part2`.
